Support regular and static operations with the same identifier

andreubotella opened the pull request:
#### 030b6006721b6d98c23068771e5d07352c89bf4c
<pre>
Support regular and static operations with the same identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=239672">https://bugs.webkit.org/show_bug.cgi?id=239672</a>

Reviewed by Darin Adler and Chris Dumez.

WebIDL used to disallow static operations from having the same name as
a regular operation of the same interface. WebKit never included this
as a check in the parser or code generator; instead, the code
generator would treat all operations with the same name as part of the
same overload set, and would produce the wrong generated code.

However, WebIDL changed to allow static and regular operations with
the same name in <a href="https://github.com/whatwg/webidl/issues/1097">https://github.com/whatwg/webidl/issues/1097</a>, which
would not share an overload set. This change implements this in
WebKit.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(LinkOverloadedOperations):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::jsTestObjPrototypeFunction_instanceAndStaticMethodWithTheSameIdentifierBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestObjConstructorFunction_instanceAndStaticMethodWithTheSameIdentifierBody):
* Source/WebCore/bindings/scripts/test/TestObj.idl:

Canonical link: <a href="https://commits.webkit.org/261472@main">https://commits.webkit.org/261472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f4232f41e460e1465903b5049b49c3fabe5fde5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/335 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11936 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3247 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99651 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104682 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45435 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13333 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/228 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13825 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9660 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19270 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52226 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7985 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15815 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM

darinadler reviewed (approved):
I probably would have preferred that we use this for something in the DOM at the same time we added it, so we have actual testing of the machinery, not just the script tests. But I won’t insist on that. The change seems fine.

andreubotella committed:
Support regular and static operations with the same identifier
https://bugs.webkit.org/show_bug.cgi?id=239672

Reviewed by Darin Adler and Chris Dumez.

WebIDL used to disallow static operations from having the same name as
a regular operation of the same interface. WebKit never included this
as a check in the parser or code generator; instead, the code
generator would treat all operations with the same name as part of the
same overload set, and would produce the wrong generated code.

However, WebIDL changed to allow static and regular operations with
the same name in https://github.com/whatwg/webidl/issues/1097, which
would not share an overload set. This change implements this in
WebKit.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(LinkOverloadedOperations):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::jsTestObjPrototypeFunction_instanceAndStaticMethodWithTheSameIdentifierBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestObjConstructorFunction_instanceAndStaticMethodWithTheSameIdentifierBody):
* Source/WebCore/bindings/scripts/test/TestObj.idl:

Canonical link: https://commits.webkit.org/261472@main
